esmvaltool/utils/recipe_filler.py

```diff
@@ -0,0 +1,684 @@
+        # iterate through directory structure looking for useful bits.
+        for base_key, fn_key in zip(basefile_split, fnfile_split):
+            if base_key == '*.nc':
+                fn_key = fn_key.replace('.nc', '')
+                start_year, end_year = fn_key.split('-')
+                output_dataset['start_year'] = start_year
+                output_dataset['end_year'] = end_year
+            elif base_key == "ensemble*.nc":
+                output_dataset['ensemble'] = fn_key
+            elif base_key == "grid*.nc":
+                output_dataset['grid'] = fn_key
+            elif base_key not in ["shortname", "ensemble*.nc", "*.nc"]:
+                output_dataset[base_key] = fn_key
+    if "exp" in file_dict:
+        if isinstance(file_dict["exp"], list):
+            output_dataset["exp"] = file_dict["exp"]
+
+    return output_dataset
+
+
+def _remove_duplicates(add_datasets):
+    """
+    Remove accidental duplicates.
+
+    Close to 0% chances this will ever be used.
+    May be used when there are actual duplicates in data
+    storage, we've seen these before, but seldom.
+    """
+    datasets = []
+    seen = set()
+
+    for dataset in add_datasets:
+        orig_exp = dataset["exp"]
+        dataset["exp"] = str(dataset["exp"])
+        tup_dat = tuple(dataset.items())
+        if tup_dat not in seen:
+            seen.add(tup_dat)
+            dataset["exp"] = orig_exp
+            datasets.append(dataset)
+
+    return datasets
+
+
+def _check_recipe(recipe_dict):
+    """Perform a quick recipe check for mandatory fields."""
+    do_exit = False
+    if "diagnostics" not in recipe_dict:
+        logger.error("Recipe missing diagnostics section.")
+        do_exit = True
+    for diag_name, diag in recipe_dict["diagnostics"].items():
+        if "variables" not in diag:
+            logger.error(f"Diagnostic {diag_name} missing variables.")
+            do_exit = True
+        for var_name, var_pars in diag["variables"].items():
+            if "mip" not in var_pars:
+                logger.error(f"Variable {var_name} missing mip.")
+                do_exit = True
+            if "start_year" not in var_pars:
+                logger.error(f"Variable {var_name} missing start_year.")
+                do_exit = True
+            if "end_year" not in var_pars:
+                logger.error(f"Variable {var_name} missing end_year.")
+                do_exit = True
+            if "exp" in var_pars:
+                if isinstance(var_pars["exp"],
+                              list) and "ensemble" not in var_pars:
+                    logger.error("Asking for experiments list for ")
+                    logger.error(f"variable {var_name} - you need to ")
+                    logger.error("define an ensemble for this case.")
+                    do_exit = True
+    if do_exit:
+        logger.error("Please fix the issues in recipe and rerun. Exiting.")
+        sys.exit(1)
+
+
+def _check_config_file(user_config_file):
```

Comment:
I'd rather see we have this functionality in `ESMValCore` and not here.

This is an excellent example of why we need the config checker we are building: https://github.com/ESMValGroup/ESMValCore/pull/868

Reply:
until then, I'd rather have it here, plus, this is slightly less restrictive than in ESMValCore, as is the case with the recipe :arrow_up:

Reply:
@stefsmeets The point of this PR is to have a first implementation and explore what is needed. If people like it, we can move the functionality to ESMValCore.